apollo-router/src/services/router_service.rs

```diff
@@ -339,61 +324,208 @@ impl RouterService {
+            Err(err) => {
+                ::tracing::error!(
+                    monotonic_counter.apollo_router_http_requests_total = 1u64,
+                    status = %err.status.as_u16(),
+                    error = %err.error,
+                );
+
+                return router::Response::error_builder()
+                    .error(
+                        graphql::Error::builder()
+                            .message(String::from("Invalid GraphQL request"))
+                            .extension_code(err.extension_code)
+                            .extension("details", err.extension_details)
+                            .build(),
+                    )
+                    .status_code(err.status)
+                    .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
+                    .context(context)
+                    .build();
+            }
+        };
+
+        let mut results = vec![];
+
+        for supergraph_request in supergraph_requests {
+            self.process_supergraph_request(&mut results, supergraph_request)
+                .await?;
+        }
+        // If we only have one result, go ahead and return it. Otherwise, create a new result
+        // which is an array of all results.
+        if results.len() == 1 {
+            Ok(results.pop().expect("should be at least one response"))
+        } else {
+            let first = results.pop().expect("we should have at least one response");
+            let (parts, body) = first.response.into_parts();
+            let context = first.context;
+            let mut array_result = "[".to_string();
+            array_result +=
+                &String::from_utf8(hyper::body::to_bytes(body).await?.to_vec()).unwrap();
+            for result in results {
+                array_result += ", ";
+                array_result += &String::from_utf8(
+                    hyper::body::to_bytes(result.response.into_body())
+                        .await?
+                        .to_vec(),
+                )
+                .unwrap();
+            }
+            array_result += "]";
+            Ok(RouterResponse {
+                response: http::Response::from_parts(parts, Body::from(array_result)),
+                context,
+            })
+        }
+    }
+
+    async fn translate_query_request(
+        &self,
+        parts: &Parts,
+    ) -> Result<Vec<graphql::Request>, TranslateError> {
+        parts.uri.query().map(|q| {
+            let mut result = vec![];
+
+            match graphql::Request::from_urlencoded_query(q.to_string()) {
+                Ok(request) => {
+                    result.push(request);
+                }
+                Err(err) => {
+                    // It may be a batch of requests, so try that (if config allows) before
+                    // erroring out
+                    if self.experimental_batching.enabled
+                        && self.experimental_batching.mode == "batch_http_link"
```

Comment:
Would be less error prone if it was an enum

Reply:
If you keep it as a `String` could you at least put this in a constant and use it everywhere @garypen ?

Reply:
done as enum: 3cc67f8da4bc72db559801292a1573b338c02484